File: web/load-fb-sdk.dart
import 'dart:html';
import "dart:async";

Future<void> loadFBSdk({required String lang}) {
  final scriptEl = ScriptElement();

  scriptEl.src =
      "https://connect.facebook.net/$lang/sdk.js#xfbml=1&version=v15.0&autoLogAppEvents=1";
  scriptEl.nonce = "XStSrfhG";

  final completer = Completer<void>();
  StreamSubscription<Event>? loadListener, errorListener;

  void onLoad(Event event) {
    loadListener?.cancel();
    errorListener?.cancel();

    completer.complete();
  }

  void onError(Event event) {
    loadListener?.cancel();
    errorListener?.cancel();
    completer.completeError(Error());
  }

  loadListener = scriptEl.onLoad.listen(onLoad);
  errorListener = scriptEl.onError.listen(onError);

  document.head?.children.add(scriptEl);

  return completer.future;
}

File: src/logic/load-fb-sdk.ts
// eslint-disable-next-line functional/no-let
let sdkLoaded = false

export function loadFBSdk(lang: string, force?: boolean) {
  if (force) {
    sdkLoaded = false
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ;(window as unknown as any).FB?.Event.clear()
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ;(window as unknown as any).FB = undefined
  }

  if (sdkLoaded) return

  return new Promise<void>((resolve, reject) => {
    const script = document.createElement("script")
    script.src = `https://connect.facebook.net/${lang}/sdk.js#xfbml=1&version=v15.0&autoLogAppEvents=1`
    script.nonce = "XStSrfhG"

    script.onload = () => {
      sdkLoaded = true
      script.onload = script.onerror = null
      resolve()
    }
    script.onerror = () => {
      script.onload = script.onerror = null
      reject()
    }
    ;(document.head || document.documentElement).appendChild(script)
    // script.remove()
  })
}
